perf: Remove reflection from NetworkBehaviour code to improve performance. [MTT-1968]

ShadauxCat opened the pull request:
## Changelog

- Changed: Improved performance of NetworkBehaviour initialization by replacing reflection when initializing NetworkVariables with compile-time code generation, which should help reduce hitching during additive scene loads. 

## Testing and Documentation

- No tests have been added... but updated an existing integration test to ensure it covered another ILPP edge case.
- No documentation changes or additions were necessary.

ShadauxCat committed:
perf: Remove reflection from NetworkBehaviour code to improve performance.

ShadauxCat committed:
changelog

0xFA11 commented on com.unity.netcode.gameobjects/CHANGELOG.md:
```diff
@@ -18,6 +18,7 @@ Additional documentation and release notes are available at [Multiplayer Documen
 - Fixed some errors that could occur if a connection is lost and the loss is detected when attempting to write to the socket. (#2495)
 
 ## Changed
+- Improved performance of NetworkBehaviour initialization, which should help reduce hitching during additive scene loads. (#2522)
```
> Improved performance of NetworkBehaviour initialization...

perhaps adding "by bla bla, which should..." might be more informative, but just nit-picking on this.

ShadauxCat replied:
Done

0xFA11 commented on com.unity.netcode.gameobjects/Editor/CodeGen/NetworkBehaviourILPP.cs:
```diff
@@ -558,6 +585,22 @@ private bool ImportReferences(ModuleDefinition moduleDefinition)
+            m_NetworkVariableBase_TypeRef = moduleDefinition.ImportReference(networkVariableBaseTypeDef);
+            foreach (var methodDef in networkVariableBaseTypeDef.Methods)
+            {
+                Console.WriteLine($"=>Method: {methodDef.Name}");
```
probably a debugging leftover :P

ShadauxCat replied:
Yep, removed, thanks for the catch.

0xFA11 commented on com.unity.netcode.gameobjects/Editor/CodeGen/NetworkBehaviourILPP.cs:
```diff
@@ -1889,6 +1946,109 @@ private void InjectWriteAndCallBlocks(MethodDefinition methodDefinition, CustomA
+                }
+                if (field.FieldType.IsSubclassOf(m_NetworkVariableBase_TypeRef))
+                {
+                    processor.Emit(OpCodes.Ldarg_0);
```
I usually find inline pseudo-C# code helpful understanding what was tried to be achieved with IL emits.

ShadauxCat replied:
Done.

NoelStephensUnity replied:
Nice... now I can understand what the OpCodes actually mean! 😸 
(TILT)

0xFA11 commented on com.unity.netcode.gameobjects/Editor/CodeGen/NetworkBehaviourILPP.cs:
```diff
@@ -1889,6 +1946,109 @@ private void InjectWriteAndCallBlocks(MethodDefinition methodDefinition, CustomA
+                initializeVariablesBaseReference = initializeVariablesBaseReference.MakeGeneric(baseTypeInstance.GenericArguments.ToArray());
+            }
+
+            processor.Emit(OpCodes.Nop);
```
^ same here

ShadauxCat replied:
Also done.

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs:
```diff
@@ -591,22 +581,7 @@ internal void InitializeVariables()
 
             m_VarInit = true;
 
-            var sortedFields = GetFieldInfoForType(GetType());
-            for (int i = 0; i < sortedFields.Length; i++)
-            {
-                var fieldType = sortedFields[i].FieldType;
-                if (fieldType.IsSubclassOf(typeof(NetworkVariableBase)))
-                {
-                    var instance = (NetworkVariableBase)sortedFields[i].GetValue(this) ?? throw new Exception($"{GetType().FullName}.{sortedFields[i].Name} cannot be null. All {nameof(NetworkVariableBase)} instances must be initialized.");
-                    instance.Initialize(this);
-
-                    var instanceNameProperty = fieldType.GetProperty(nameof(NetworkVariableBase.Name));
-                    var sanitizedVariableName = sortedFields[i].Name.Replace("<", string.Empty).Replace(">k__BackingField", string.Empty);
-                    instanceNameProperty?.SetValue(instance, sanitizedVariableName);
-
-                    NetworkVariableFields.Add(instance);
-                }
-            }
```
what this code was doing anyway? why did we try to access variable name?

ShadauxCat replied:
Based on the tests that failed when I forgot to bring over the "k__BackingField" part of this, it seems that the network variable's name is used for tools/metrics. This is just setting a runtime property to match the variable's name in the source, so basically it's doing `Foo.Name = "Foo"`.

ShadauxCat committed:
- Added some comments to clarify some ILPP code
- Added an extra error handler to catch an edge case I missed before
- Added a bit more info to the changelog
- Fixed testproject-tools-integration tests

NoelStephensUnity reviewed (approved):
Looks good to me!
@ThusSpokeNomad ?

0xFA11 commented on com.unity.netcode.gameobjects/Editor/CodeGen/NetworkBehaviourILPP.cs:
```diff
@@ -785,6 +827,18 @@ private bool ImportReferences(ModuleDefinition moduleDefinition)
                 }
             }
 
+            // Standard stuff is really hard to reliably find using the normal way
+            // It's different in mono vs. .net core
+            // Importing with typeof() is less dangeroud for standard stuff though
+            // so we can just do it
```
```suggestion
            // Standard types are really hard to reliably find using the Mono Cecil way, they resolve differently in Mono vs .NET Core
            // Importing with typeof() is less dangerous for standard framework types though, so we can just do it
```

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs:
```diff
@@ -551,35 +549,27 @@ internal void InternalOnLostOwnership()
-            }
-
-            return s_FieldTypes[type];
+            // ILPP generates code for all NetworkBehaviour subtypes to initialize
+            // each type's network variables.
```
```suggestion
            // ILPP generates code for all NetworkBehaviour subtypes to initialize each type's network variables.
```

We don't have a max line length limit. Let's keep sentences unbroken?

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs:
```diff
@@ -551,35 +549,27 @@ internal void InternalOnLostOwnership()
-        private static FieldInfo[] GetFieldInfoForTypeRecursive(Type type, List<FieldInfo> list = null)
+#pragma warning disable IDE1006 // disable naming rule violation check
+        // RuntimeAccessModifiersILPP will make this `protected`
+        // Using this method here because ILPP doesn't seem to let us do visibility
+        // modification on properties.
```
```suggestion
        // Using this method here because ILPP doesn't seem to let us do visibility modification on properties.
```

^

0xFA11 commented on com.unity.netcode.gameobjects/Editor/CodeGen/NetworkBehaviourILPP.cs:
```diff
@@ -1889,6 +1945,133 @@ private void InjectWriteAndCallBlocks(MethodDefinition methodDefinition, CustomA
+            {
+                if (methodDefinition.Name == k_NetworkBehaviour___initializeVariables)
+                {
+                    // If this hits, we've already generated the method for this class
+                    // because a child class got processed first.
```
```suggestion
                    // If this hits, we've already generated the method for this class because a child class got processed first.
```

^

0xFA11 reviewed (approved):
Added a few suggestions for inline comments (typo, clarity, style).
But other than that, 🚀

ShadauxCat committed:
Apply suggestions from code review

Co-authored-by: Fatih Mar <mfatihmar@gmail.com>

netcode-ci-service committed:
Merge develop into perf/remove_reflection_in_networkbehaviour
